### DISCUSSION TITLE
Refactoring suggestion

### DISCUSSION THREAD

**Discussion by ITEAD-Thierry** [ITEAD-Thierry/NextionX#11] [Ideas]
@sstaub from Munich/Germany made an interesting post and suggestions yesterday by opening an issue on this repository. Since such is not an issue, I opened the discussion here by citing him as follows:

**_Thanks for the inspiring library and the Blog. I found many things in the code which might done better, so I worked on it.
Better variable names, Doxygen style documentation, functions which can deleted, and deleting most of the this-> stuff which is not needed ...
I tried to separate the .h file into a .h and a .cpp file but with no luck, it seems this a problem with templates I have to investigate.
Feel free to overtake (some) changes.
The refactored lib is attached, with an example for Mega.
NextionX.zip_**

**Comment by ITEAD-Thierry** [comment]
Thank you very much for that! I looked quickly over your variant - it looks promising. You created doxygen tags for methods which are only temporarily public in this interim release but which are intended to be protected or private in later releases. That’s why in my manual documentation, some are intentionally not mentioned. I have in the meantime added (but not yet fully tested and thus not yet published) a bunch of new functions to read attributes back from Nextion components into the MCU, asynchronously (not blocking), so I can consider your optimizations only at a later moment.
Although I admire everything which is made better as my code, we have to respect the multi-platform compatibility, everything has to compile in C++11 (for Arduino), C++14 (for Teensy), and C++17 (for ESP). So, testing and integrating your mods might come only in a few weeks.
Template parameters require always that declarations and definitions are in the same .h file.

**Comment by ITEAD-Thierry** [comment]
Here I cite again @sstaub‘s reply:

**_Debug works, I renamed most of the functions and variables. e.g. using `_var` is old C style, also Set and Get methods are no more needed when using overloading functions. I wrote the very popular Ticker library which works on all platforms. The problem I see is using the Arduino String library which is a problem for microcontrollers with small RAM, cause of defragmentation of the heap.
With non blocking did you mean interrupt based? Receiving data from the Nextion display with RX interrupts will solve many timing problems._**

**Comment by ITEAD-Thierry** [comment]
With non blocking, I mean launching another custom callback function of a NexComp when the requested answer has fully arrived.
I’m aware of the fact that many things can be done in a more modern style, but Arduino users, especially beginners, seem to feel more comfortable with familiar naming conventions and the good old getSomething, setSomething, writeSomething and readSomething methods.

Naturally, one can eliminate set and get functions and assign object members directly, but it remains much more secure to keep the latter private and access these through public functions which add an opportunity for additional validation.

BTW: Removing the this-> stuff doesn‘t give any advantage, neither for the compiled bin or hex file size, nor for execution speed, since the compiler will use an implicit this-> in such cases. But without written this->, it’s more difficult to see if you are acting on a local variable/function or an object member/method.

As promised, I will have a closer look to that, though, in a few weeks when I‘ll have more free time and less overload at my day job.

**Reply by sstaub** [reply]
`this->` is  needed to avoid conflict with a class variable name and variable name of functions, so the classic C style with `_var` is no more needed when using C++, that is the way I use `this->`
In your code only four times `this->`and two real `this` as a reference  are needed.
General avoiding templates is helpful for beginners, I think templates become obsolete with C++20, maybe excepting class templates, I'm not going further in. 

> Template parameters require always that declarations and definitions are in the same .h file.

Thank you for this advice, that's why avoid templates.

**Reply by ITEAD-Thierry** [reply]
I‘m a huge fan of single header libraries and of templates because both allow (and sometimes force) you to keep the source code small and extremely compact. 
I guess that it‘s a matter of taste and habitude. 
Templates are helpful to implement a function only once, even for different parameter types, which makes code easier to maintain. Even if this might change in C++20, it‘s still a long way, the current Arduino toolchain is still on C++11…
I agree with you on the Arduino String class and its weak points, but after finding out that there is no printf() support in Arduino to assemble or concat variable length char arrays, only sprintf() which requires additional char buffers and which increases the compiled file size by 2kB, I don‘t see an alternative.

**Reply by sstaub** [reply]
Putting all things in one .h file is ok for small code, but when the project grows, it becomes a problem. 
Instead of String you can also use the C Strings byte library , it is not so comfortable but works. e.g. [https://github.com/sstaub/OSC](https://github.com/sstaub/OSC)
Therefore it will good to know if there is a limit in the length of the commands sending to the display?

**Reply by ITEAD-Thierry** [reply]
There is a theoretical limit of 65530 characters when getting or setting the txt attribute of a Nextion Text component...

There is one of your modifications which I can't adopt: You changed the event parameter type of the NexComp::callback function from uint8_t to bool. But more callback types than only touch and release are coming, for example to notify the completion of an asynchronous getAttr() request.

**Reply by sstaub** [reply]
Yes, the change to bool is mistake from me, I realized also that they're coming more callback types. Maybe some kind of enumeration of callback types will helpful. Waiting for your next release.

**Comment by sstaub** [comment]
Attached a file with elimination of all Arduino `String`

[NextionX.h.zip](https://github.com/ITEAD-Thierry/NextionX/files/6085211/NextionX.h.zip)
